### PR TITLE
refactor: Contacts menu to Vue

### LINKED ISSUE #15586: Vue contacts menu
**Is your feature request related to a problem? Please describe.**

The contacts menu is plain old jQuery.

**Describe the solution you'd like**

I would like to replace that with a Vue component.

As per discussion below the contacts menu should actually get merged into the unified search: https://github.com/nextcloud/server/issues/15586#issuecomment-901268383

**Describe alternatives you've considered**

n/a

**Additional context**

https://github.com/orgs/nextcloud/projects/31

---

@nextcloud/javascript @nextcloud/vue objections?

### LINKED ISSUE #39772: [Bug]: Contacts menu popover overflow is not visible
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

The contact popover menu in the top-right contacts menu can be too wide and the overflow is not visible.

image

### Steps to reproduce

1. Create a user with a long first and last name
2. Install Talk (spreed)
3. Open the top-right header contacts menu
4. Search for the user
5. Open the "..." menu

### Expected behavior

The overflow could be visible or the popover could have a max width that would prevent it to overflow.

### Installation method

None

### Nextcloud Server version

master

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.1

### Web server

Apache (supported)

### Database engine version

SQlite

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by ChristophWurst**
* Fixes https://github.com/nextcloud/server/issues/15586
* Fixes https://github.com/nextcloud/server/issues/39772
* Enables https://github.com/nextcloud/server/issues/40559

## Summary

The contacts menu is still based on jQuery, Backbone and handlebars.

### Changes

* The *top action* is inline, the rest is moved into an actions menu
* Small styling changes due to move to standardized Vue components

### Screenshots

| Before | After |
|--------|--------|
| Bildschirmfoto vom 2023-10-03 15-10-16 |Bildschirmfoto vom 2023-10-03 15-17-43 | 

## TODO

- [x] Migrate by moving code around
- [x] Fix search
- [x] Migrate code to BEM
- [x] ~~Replace icon classes with material icons~~ not possible because icon URIs are provided via public API

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by ChristophWurst** [comment]
Requesting early feedback. I'll try to organize the styles into BEM. Otherwise this should be mergeable.

File: core/src/components/ContactsMenu/Contact.vue [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		</a>
+		<NcAvatar v-else-if="contact.avatar"
+			class="avatar"
+			:is-no-user="true"
+			:display-name="contact.avatarLabel"
+			:url="contact.avatar" />
+		<NcAvatar v-else
+			class="avatar"
+			:is-no-user="true"
+			:display-name="contact.avatarLabel" />
```
Comment (skjnldsv) [inline core/src/components/ContactsMenu/Contact.vue:47]:
```suggestion
		<NcAvatar
			class="avatar"
			:is-no-user="true"
			:display-name="contact.avatarLabel"
			:url="contact.avatar || undefined" />
```

Wouldn't this work ?

**Review by skjnldsv — COMMENTED** [review]
Amazing!! :crown: 

Some rules to consider though:
- [x] Classes are a bit messy, sometimes it looks like BEM, sometimes it doesn't. Let's be compliant ad follow BEM entirely? But maybe you just reused the old css classes and wanted to keep changes to a minimum, I don't know)
- [x] `icon-loading` should be replaced with Vue `NcLoadingIcon` component
- [x] `emptycontent` class style should be replaced with `NcEmptyContent` component
- [x] Variables should have strict typing if possible (like `this.loadingText`)
- [x] Maybe reset the error when searching again? (See `this.error`)
- [x] `contactsAppEnabled` is always `true`
- [x] ~~`searchTerm` should reset on close maybe? Not sure :)~~

File: core/src/views/ContactsMenu.vue [hunk 2]
```diff
@@ -27,41 +27,91 @@
...
-	},
-
 	methods: {
-		handleOpen() {
-			this.contactsMenu?.loadContacts()
+		async handleOpen() {
+			await this.getContacts('')
+		},
+		async getContacts(searchTerm) {
+			this.loadingText = t('core', 'Loading your contacts …')
```
Comment (skjnldsv) [inline core/src/views/ContactsMenu.vue:101]: Reset the error when searching again?

**Comment by ChristophWurst** [comment]
> Classes are a bit messy, sometimes it looks like BEM, sometimes it doesn't. Let's be compliant ad follow BEM entirely? But maybe you just reused the old css classes and wanted to keep changes to a minimum, I don't know)

See PR description

**Comment by ChristophWurst** [comment]
> `searchTerm` should reset on close maybe? Not sure :)

I did not reset before and unified search doesn't do that either. I'm keeping the search term.

**Comment by ChristophWurst** [comment]
@skjnldsv I can add Vue component unit tests as a replacement for the removed js specs. Would you be okay with that? I'd install and use https://www.npmjs.com/package/@vue/test-utils on top of jest.

**Comment by skjnldsv** [comment]
@ChristophWurst absolutely!
You can also do cypress component testing if you prefer, we have multiple options now :)

**Comment by ChristophWurst** [comment]
image

Moved a few more pixels to get close to the old appearance. This is as good as it gets with my frontend skills.

**Comment by ChristophWurst** [comment]
Branch updated to latest master and removed two accidentally added files.

@skjnldsv please double-check if I fulfilled your requested changes.

**Comment by ChristophWurst** [comment]
@skjnldsv @Pytal friendly ping. I need this to start https://github.com/nextcloud/server/issues/40559.

**Review by skjnldsv — APPROVED** [review]
Only complain I have is the vertical centering of the avatar, but all good otherwise! :rocket: 
image

EDIT: ah, and the roundness of this icon:
image

**Review by Pytal — APPROVED** [review]
🙌

**Comment by Pytal** [comment]
Failing CI related

**Comment by skjnldsv** [comment]
> @skjnldsv did you compile the sources? 🤭
> 
> Because those two bugs are fixed with the rework:

I think I did, but maybe I did it wrong :see_no_evil:

**Comment by ChristophWurst** [comment]
Pushed a rebased, squashed and compiled version.

**Comment by szaimen** [comment]
CI failure related afaics

**Comment by ChristophWurst** [comment]
as green as it gets (red)

acceptance-headers smells related :face_exhaling:

**Comment by szaimen** [comment]
@ChristophWurst how much work is still to be done here? Asking because of https://github.com/nextcloud/server/pull/40873

**Comment by ChristophWurst** [comment]
``UsersSettingsContext::iCreateUserWithPassword()`` doesn't work anymore. There is an open modal afterwards, the user is not created. Acceptance tests that create a second user fail.

**Comment by szaimen** [comment]
🎉🎉🎉🎉🎉

**Pull request merged**